clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java

```diff
@@ -94,53 +106,48 @@ private boolean needsNewToken() {
-                signedRequest.getSignature(),
-                Arrays.asList(signedRequest.getSignedHeadersParam().split(";")),
-                signedRequest.getVersion(),
-                signedRequest.getAlgorithm(),
-                signedRequest.getSecurityToken()
-        );
+        LakeFSExternalPrincipalIdentityRequest identityTokenParams = new LakeFSExternalPrincipalIdentityRequest(signedRequest.getHTTPMethod(), signedRequest.getHost(), signedRequest.getRegion(), signedRequest.getAction(), signedRequest.getDate(), signedRequest.getExpires(), signedRequest.getAccessKeyId(), signedRequest.getSignature(), Arrays.asList(signedRequest.getSignedHeadersParam().split(";")), signedRequest.getVersion(), signedRequest.getAlgorithm(), signedRequest.getSecurityToken());
+
+        // base64 encode
+        return Base64.encodeBase64URLSafeString(identityTokenParams.toJSON().getBytes());
+    }
+
+    public String newPresignedGetCallerIdentityTokenRequest() throws Exception {
+        GeneratePresignGetCallerIdentityResponse signedRequest = this.newPresignedRequest();
+
+        // generate token parameters object
+        LakeFSExternalPrincipalIdentityRequest identityTokenParams = new LakeFSExternalPrincipalIdentityRequest(signedRequest.getHTTPMethod(), signedRequest.getHost(), signedRequest.getRegion(), signedRequest.getAction(), signedRequest.getDate(), signedRequest.getExpires(), signedRequest.getAccessKeyId(), signedRequest.getSignature(), Arrays.asList(signedRequest.getSignedHeadersParam().split(";")), signedRequest.getVersion(), signedRequest.getAlgorithm(), signedRequest.getSecurityToken());
 
         // base64 encode
-        return Base64.encodeBase64String(identityTokenParams.toJSON().getBytes());
+        return Base64.encodeBase64URLSafeString(identityTokenParams.toJSON().getBytes());
     }
 
     private void newToken() throws Exception {
+        // created identity token to exchange for lakeFS token
         String identityToken = this.newPresignedGetCallerIdentityToken();
-        /*
-        TODO(isan)
-         depends on missing functionality PR https://github.com/treeverse/lakeFS/pull/7578 being merged.
-         before merging this code - implement the call to lakeFS.
-         it will introduce the functionality in the generated client of actually doing the login.
-         call lakeFS to exchange the token for a lakeFS token
-         The flow will be:
-         1. use this.lakeFSApi Client with ExternalPrincipal API class (no auth required)
-         2. this.lakeFSAuthToken = call api.ExternalPrincipalLogin(identityToken, <lakeFS Token optional TTL>)
-        */
-        // dummy initiation
-        this.lakeFSAuthToken = new AuthenticationToken();
-        this.lakeFSAuthToken.setTokenExpiration(System.currentTimeMillis() + 60);
+
+        // build lakeFS login request
+        ExternalLoginInformation req = new ExternalLoginInformation();
+
+        if (this.lakeFSTokenTTLSeconds.isPresent()) {
```

Comment:
Could use [this](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#flatMap-java.util.function.Function-) to apply an operator directly if present.